pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,368 @@
+	}
+
+	// check destinations
+	if !destinationsMatch(resolution.destinations, placementDecisionSpec.Destinations) {
+		return false
+	}
+
+	// check workload
+	return workloadMatchesPlacementDecisionSpec(&placementDecisionSpec.Workload, resolution.objectIdentifierToKey, gvkGvrMapper)
+}
+
+// destinationsMatch returns true if the destinations in the resolution
+// match the destinations in the placement decision spec.
+// The function performs one pass over both slices, sequentially.
+func destinationsMatch(resolvedDestinations sets.Set[string], placementDecisionDestinations []v1alpha1.Destination) bool {
+	if len(resolvedDestinations) != len(placementDecisionDestinations) {
+		return false
+	}
+
+	for _, destination := range placementDecisionDestinations {
+		if !resolvedDestinations.Has(destination.ClusterId) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// workloadMatchesPlacementDecisionSpec returns true if the workload in the
+// resolution matches the workload in the placement decision spec.
+func workloadMatchesPlacementDecisionSpec(placementDecisionSpecWorkload *v1alpha1.DownsyncObjectReferences,
+	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
+	// check lengths
+	clusterScopedObjectsCount := 0
+	for _, clusterScopeDownsyncObjects := range placementDecisionSpecWorkload.ClusterScope {
+		clusterScopedObjectsCount += len(clusterScopeDownsyncObjects.ObjectNames)
+	}
+
+	namespacedObjectsCount := 0
+	for _, namespaceScopeDownsyncObjects := range placementDecisionSpecWorkload.NamespaceScope {
+		for _, objectsByNamespace := range namespaceScopeDownsyncObjects.ObjectsByNamespace {
+			namespacedObjectsCount += len(objectsByNamespace.Names)
+		}
+	}
+
+	if len(objectIdentifierToKeyMap) != clusterScopedObjectsCount+namespacedObjectsCount {
+		return false
+	}
+
+	// again we can check match by making sure all objects are mapped, since entries are unique and the length is equal.
+	// check cluster-scoped all exist
+	if !placementDecisionClusterScopeIsMapped(placementDecisionSpecWorkload.ClusterScope, objectIdentifierToKeyMap,
+		gvkGvrMapper) {
+		return false
+	}
+
+	// check namespace-scoped all exist
+	return namespaceScopeMatchesPlacementDecisionSpec(placementDecisionSpecWorkload.NamespaceScope, objectIdentifierToKeyMap,
+		gvkGvrMapper)
+}
+
+// clusterScopeMatchesPlacementDecisionSpec returns true if the cluster-scope
+// section in the placement decision spec all exist in the resolution.
+func placementDecisionClusterScopeIsMapped(placementDecisionSpecClusterScope []v1alpha1.ClusterScopeDownsyncObjects,
+	objectIdentifierToKeyMap map[string]*util.Key, gvkGvrMapper util.GvkGvrMapper) bool {
+	for _, clusterScopeDownsyncObjects := range placementDecisionSpecClusterScope {
+		gvr := schema.GroupVersionResource(clusterScopeDownsyncObjects.GroupVersionResource)
+		gvk, found := gvkGvrMapper.GetGvk(gvr)
+
+		if !found {
+			return false // if not found then not mapped and not in resolution
```

Comment:
Or maybe there was a match when the relevant objects were read but the relevant CRD and all objects whose resource (Kind) was defined by it have been deleted since then and the CRD deletion has been processed by this controller but the object's key is still in the resolution's map (set) of object keys because there is no synchronization that would prevent that.

Reply:
Good point, this adds more backing to why the key should be reworked (or at least the key I use in the resolution). This issue would be resolved when the said rework is done, therefore I'll leave it as-is for now, ok?

Reply:
OK